framework/decode/preload_file_processor.cpp

```diff
@@ -0,0 +1,369 @@
+                            }
+                        }
+                        else
+                        {
+                            bool should_break = false;
+                            success           = ProcessMethodCall(block_header, api_call_id, should_break);
+                            if (should_break)
+                            {
+                                break;
+                            }
+                        }
+                    }
+                    else
+                    {
+                        HandleBlockReadError(kErrorReadingBlockHeader, "Failed to read function call block header");
+                    }
+                }
+                else if (format::RemoveCompressedBlockBit(block_header.type) == format::BlockType::kMetaDataBlock)
+                {
+                    if (status_ == PreloadStatus::kRecord)
+                    {
+                        preload_buffer_.Reserve(sizeof(block_header) + block_header.size);
+                        preload_buffer_.Add(&block_header);
+                        size_t parameters_size  = block_header.size;
+                        auto*  parameter_buffer = preload_buffer_.Add(parameters_size);
+                        success                 = ReadBytes(parameter_buffer, parameters_size);
+                        if (!success)
+                        {
+                            HandleBlockReadError(kErrorReadingBlockData, "Failed to preload meta-data block");
+                        }
+                    }
+                    else
+                    {
+                        format::MetaDataId meta_data_id = format::MakeMetaDataId(
+                            format::ApiFamilyId::ApiFamily_None, format::MetaDataType::kUnknownMetaDataType);
+
+                        success = ReadBytes(&meta_data_id, sizeof(meta_data_id));
+
+                        if (success)
+                        {
+                            success = ProcessMetaData(block_header, meta_data_id);
+                        }
+                        else
+                        {
+                            HandleBlockReadError(kErrorReadingBlockHeader, "Failed to read meta-data block header");
+                        }
+                    }
+                }
+                else if (block_header.type == format::BlockType::kFrameMarkerBlock)
+                {
+                    format::MarkerType marker_type  = format::MarkerType::kUnknownMarker;
+                    uint64_t           frame_number = 0;
+
+                    success = ReadBytes(&marker_type, sizeof(marker_type));
+
+                    if (success)
+                    {
+                        if (status_ == PreloadStatus::kRecord)
+                        {
+                            const auto is_frame_delimiter = IsFrameDelimiter(block_header.type, marker_type);
+                            preload_buffer_.Reserve(sizeof(block_header) + block_header.size);
```

Comment:
I almost think these blocks should be in a helper function like:   `bool ReadPreloadBuffer(size_t header_size, size_t ignore_size)` or something like that.  Then you only have to pass in the things like `sizeof(marker_type)` here and appropriate items above .   And it will make this cleaner.

Reply:
I concur, I will add that and refactor code to use it.